impeller/entity/entity_pass.cc

```diff
@@ -891,21 +903,75 @@ bool EntityPass::OnRender(
                   renderer, clip_coverage_stack, global_pass_position);
   }
 
-  bool is_collapsing_clear_colors = !collapsed_parent_pass &&
-                                    // Backdrop filters act as a entity before
-                                    // everything and disrupt the optimization.
-                                    !backdrop_filter_proc_;
-  for (const auto& element : elements_) {
-    // Skip elements that are incorporated into the clear color.
-    if (is_collapsing_clear_colors) {
-      auto [entity_color, _] =
-          ElementAsBackgroundColor(element, clear_color_size);
-      if (entity_color.has_value()) {
-        continue;
+  bool should_collapse_clear_colors =
+      !collapsed_parent_pass &&
+      // Backdrop filters act as a entity before
+      // everything and disrupt the optimization.
+      !backdrop_filter_proc_;
+
+  // Count the number of elements eaten by the clear color optimization.
+  size_t opaque_clear_entity_count = 0;
+  size_t translucent_clear_entity_count = 0;
+  if (should_collapse_clear_colors) {
+    for (const auto& element : elements_) {
+      if (const Entity* entity = std::get_if<Entity>(&element)) {
+        std::optional<Color> entity_color =
+            entity->AsBackgroundColor(clear_color_size);
+        if (entity_color.has_value()) {
+          if (entity->GetBlendMode() == BlendMode::kSource) {
+            opaque_clear_entity_count++;
+          } else {
+            translucent_clear_entity_count++;
+          }
+          continue;
+        }
+        break;
       }
-      is_collapsing_clear_colors = false;
     }
+  }
+
+  using ElementCallback = std::function<bool(const Element&)>;
+  using ElementIterator = std::function<bool(const ElementCallback&)>;
+
+  ElementIterator element_iterator;
+
+  if (renderer.GetDeviceCapabilities().SupportsFramebufferFetch()) {
+    element_iterator =
```

Comment:
I think this should be fairly straightforward to port to experimental canvas, but we'd need to be able to configure the dispatch behavior.

We'd need to dispatch (per save Layer) once backwards through the draws for backdrop independent elements, and once forward pass for backdrop dependent elements,

FYI @flar

Reply:
The storage buffer is not bidirectionally traversable for a couple of reasons

- it is packed rather than indexed with offsets between elements
- if we created an index, much of the state data is recorded incrementally
  - this includes transform, clips, paint attributes

The entities are not incremental for any state, they hold all of their state in absolute terms so it is easy to go forwards/backwards or even to just visit a single entity in isolation.

There is a bit of redesign that would have to happen to make that work.

Reply:
bidirection iteration isn't strictly necessary, another approach would be to only iterate forwards and record the internal offsets, then allow redispatching a subset of them.

Reply:
I should have said "... bidirectional, or even random access dispatching ...". Note the second bullet point starts with "If we created an index" meaning that isn't enough to get us where we need to be for this technique...

Reply:
We might not want to expose that as "here you go save some pointers and offsets", but since the recorded structure is immutable-ish I'd imagine there is a safe interface we could build for that?

Reply:
We might be able to get away with a list of structures containing:

- buffer offset of op
- fully populated Paint representing the attribute state at that offset
- impeller::Matrix representing the transform at that offset
- (clip stuff??)

Reply:
I'm not sure what clip stuff we'd need. Probably whatever the DrawOrder optimizer here is recording.